From 6763260a5f8de825cb3fa7802b291ff3bca37628 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Fri, 16 Feb 2024 15:58:22 +0100
Subject: [PATCH 0606/1695] drivers: stm32_tamp: support STM32MP23x SoC family

Add support for STM32MP23 SoC family.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 411413b75745980bf5d19b0d63021f67d4f8c497 drivers: stm32_tamp: support STM32MP23x SoC family]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: Iaaf1c4c603ccc97f3a3ee7a8c83ae08a53807514
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/360738
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/stm32_tamp.c | 34 +++++++++++++++++++++++++++++-----
 1 file changed, 29 insertions(+), 5 deletions(-)

diff --git a/core/drivers/stm32_tamp.c b/core/drivers/stm32_tamp.c
index 0c98119df..ca54ccdf9 100644
--- a/core/drivers/stm32_tamp.c
+++ b/core/drivers/stm32_tamp.c
@@ -21,9 +21,9 @@
 #include <libfdt.h>
 #include <mm/core_memprot.h>
 #include <stdbool.h>
-#ifdef CFG_STM32MP25
+#if defined(CFG_STM32MP25) || defined(CFG_STM32MP23)
 #include <stm32_sysconf.h>
-#endif /* CFG_STM32MP25 */
+#endif /* defined(CFG_STM32MP25) || defined(CFG_STM32MP23) */
 #include <stm32_util.h>
 
 
@@ -302,7 +302,7 @@ static const char * const itamper_name[] = {
 DECLARE_KEEP_PAGER(itamper_name);
 #endif
 
-#ifdef CFG_STM32MP25
+#if defined(CFG_STM32MP25) || defined(CFG_STM32MP23)
 static const char * const itamper_name[] = {
 	[INT_TAMP1] = "Backup domain voltage threshold monitoring",
 	[INT_TAMP2] = "Temperature monitoring",
@@ -359,6 +359,12 @@ static struct stm32_tamp_conf ext_tamp_mp25[] = {
 	{ .id = EXT_TAMP7 }, { .id = EXT_TAMP8 },
 };
 
+static struct stm32_tamp_conf ext_tamp_mp23[] = {
+	{ .id = EXT_TAMP1 }, { .id = EXT_TAMP2 }, { .id = EXT_TAMP3 },
+	{ .id = EXT_TAMP4 }, { .id = EXT_TAMP5 }, { .id = EXT_TAMP6 },
+	{ .id = EXT_TAMP7 },
+};
+
 #define GPIO_BANK(port)	 ((port) - 'A')
 
 static const struct stm32_tamp_pin_map pin_map_mp13[] = {
@@ -1980,10 +1986,10 @@ static TEE_Result stm32_tamp_probe(const void *fdt, int node,
 		goto err;
 	}
 
-#ifdef CFG_STM32MP25
+#if defined(CFG_STM32MP25) || defined(CFG_STM32MP23)
 	if (stm32_tamp.pdata.mask_pot_reset)
 		stm32mp_syscfg_write(SYSCFG_POTTAMPRSTCR, BIT(0), BIT(0));
-#endif /* CFG_STM32MP25 */
+#endif /* defined(CFG_STM32MP25) || defined(CFG_STM32MP23) */
 
 	/*
 	 * Select extra IP to add in the deleted/blocked IP in case of
@@ -2126,8 +2132,26 @@ static const struct stm32_tamp_compat mp25_compat = {
 		.pin_map_size = ARRAY_SIZE(pin_map_mp25),
 };
 
+static const struct stm32_tamp_compat mp23_compat = {
+		.nb_monotonic_counter = 2,
+		.tags = TAMP_HAS_REGISTER_SECCFGR |
+			TAMP_HAS_REGISTER_PRIVCFGR |
+			TAMP_HAS_RIF_SUPPORT |
+			TAMP_HAS_REGISTER_ERCFGR |
+			TAMP_HAS_REGISTER_CR3 |
+			TAMP_HAS_REGISTER_ATCR2 |
+			TAMP_HAS_CR2_SECRET_STATUS,
+		.int_tamp = int_tamp_mp25,
+		.int_tamp_size = ARRAY_SIZE(int_tamp_mp25),
+		.ext_tamp = ext_tamp_mp23,
+		.ext_tamp_size = ARRAY_SIZE(ext_tamp_mp23),
+		.pin_map = pin_map_mp25,
+		.pin_map_size = ARRAY_SIZE(pin_map_mp25),
+};
+
 static const struct dt_device_match stm32_tamp_match_table[] = {
 	{ .compatible = "st,stm32mp25-tamp", .compat_data = &mp25_compat },
+	{ .compatible = "st,stm32mp23-tamp", .compat_data = &mp23_compat },
 	{ .compatible = "st,stm32mp13-tamp", .compat_data = &mp13_compat },
 	{ .compatible = "st,stm32-tamp", .compat_data = &mp15_compat },
 	{ }
-- 
2.47.2

